assets/src/setup/components/reader-themes-context-provider.js

```diff
@@ -0,0 +1,155 @@
+/**
+ * WordPress dependencies
+ */
+import { createContext, useEffect, useState, useRef, useContext, useMemo } from '@wordpress/element';
+import apiFetch from '@wordpress/api-fetch';
+
+/**
+ * External dependencies
+ */
+import PropTypes from 'prop-types';
+/**
+ * Internal dependencies
+ */
+import { Options } from './options-context-provider';
+
+export const ReaderThemes = createContext();
+
+/**
+ * Context provider for options retrieval and updating.
+ *
+ * @param {Object} props Component props.
+ * @param {string} props.ajaxurl WP AJAX URL.
+ * @param {?any} props.children Component children.
+ * @param {string} props.readerThemesEndpoint REST endpoint to fetch reader themes.
+ * @param {string} props.updatesNonce Nonce for the AJAX request to install a theme.
+ */
+export function ReaderThemesContextProvider( { ajaxurl, children, readerThemesEndpoint, updatesNonce } ) {
+	const [ themes, setThemes ] = useState( null );
+	const [ fetchingThemes, setFetchingThemes ] = useState( false );
+	const [ themeFetchError, setThemeFetchError ] = useState( null );
+	const [ downloadingTheme, setDownloadingTheme ] = useState( false );
+	const [ downloadingThemeError, setDownloadingThemeError ] = useState( null );
+
+	const { options, savingOptions } = useContext( Options );
+	const { reader_theme: readerTheme } = options || {};
+
+	// This component sets state inside async functions. Use this ref to prevent state updates after unmount.
+	const hasUnmounted = useRef( false );
+
+	const selectedTheme = useMemo(
+		() => themes ? themes.find( ( { slug } ) => slug === readerTheme ) : null,
+		[ readerTheme, themes ],
+	);
+
+	/**
+	 * Downloads the selected reader theme, if necessary, when options are saved.
+	 */
+	useEffect( () => {
+		if ( ! selectedTheme ) {
+			return;
+		}
+
+		if ( ! savingOptions || downloadingTheme ) {
+			return;
+		}
+
+		if ( 'installable' !== selectedTheme.availability ) {
+			return;
+		}
+
+		/**
+		 * Downloads a theme from WordPress.org using the traditional AJAX action.
+		 */
+		async function downloadReaderTheme() {
+			setDownloadingTheme( true );
+
+			try {
+				const body = new global.FormData();
+				body.append( 'action', 'install-theme' );
+				body.append( 'slug', selectedTheme.slug );
+				body.append( '_wpnonce', updatesNonce );
+
+				// This is the only fetch request in the setup wizard that doesn't go to a REST endpoint.
+				// We need to use window.fetch to bypass the apiFetch middlewares that are useful for other requests.
+				await global.fetch( ajaxurl, {
+					body,
+					method: 'POST',
+				} );
+
+				if ( true === hasUnmounted.current ) {
+					return;
+				}
+			} catch ( e ) {
+				if ( true === hasUnmounted.current ) {
+					return;
+				}
+
+				setDownloadingThemeError( e );
+			}
+
+			setDownloadingTheme( false );
+		}
+
+		downloadReaderTheme();
```

Comment:
Is there a particular reason why this is not an anonymous IIA(A)FE?

I.e.:
```suggestion
		( async () => {
			setDownloadingTheme( true );

			try {
				const body = new global.FormData();
				body.append( 'action', 'install-theme' );
				body.append( 'slug', selectedTheme.slug );
				body.append( '_wpnonce', updatesNonce );

				// This is the only fetch request in the setup wizard that doesn't go to a REST endpoint.
				// We need to use window.fetch to bypass the apiFetch middlewares that are useful for other requests.
				await global.fetch( ajaxurl, {
					body,
					method: 'POST',
				} );

				if ( true === hasUnmounted.current ) {
					return;
				}
			} catch ( e ) {
				if ( true === hasUnmounted.current ) {
					return;
				}

				setDownloadingThemeError( e );
			}

			setDownloadingTheme( false );
		} )();
```

Reply:
Ok, the suggestion diff is pretty silly, I basically just wrapped the thing in parentheses... 😄

Reply:
I saw this on a different project recently and kinda like it, especially if you have all the conditions for bailing early at the top of the effect callback. I'm going to make this update.